### PR TITLE
Adds option to allow Kueue/Jobset to be installed on a GKE cluster via blueprints

### PR DESCRIPTION
**Pull request by ankitkinra**
Allows blueprint to install Kueue/Jobset for a GKE cluster which reduces the manual steps required by the user to configure these frameworks.

In this PR:
* Downloaded versions for kueue / jobset and updated the manifest to add tolerations for the system pool taint.

Testing 
* Manually tested that a3 high cluster comes up with kueue and jobset controllers

Example yaml in a blueprint

```yaml
- id: workload_component_install
    source: modules/management/kubectl-apply
    use: [gke_cluster]
    settings:
      kueue:
        install: true
        config_path: $(ghpc_stage("manifests/user-provided-kueue-config.yaml"))
      jobset:
        install: true
```

Future enhancements 
* Automatic basic configuration for Kueue

### Submission Checklist

Please take the following actions before submitting this pull request.

* Fork your PR branch from the Toolkit "develop" branch (not main)
* Test all changes with pre-commit in a local branch [#](https://goo.gle/hpc-toolkit#development)
* Confirm that "make tests" passes all tests
* Add or modify unit tests to cover code changes
* Ensure that unit test coverage remains above 80%
* Update all applicable documentation
* Follow Cluster Toolkit Contribution guidelines [#](https://goo.gle/hpc-toolkit-contributing)

### CODE REVIEW

File: examples/gke-a3-highgpu.yaml
Comment (nick-stroud) [inline examples/gke-a3-highgpu.yaml:62]: I suspect this will be problematic and in edge cases we will end up with workloads getting placed on the system pool. Instead of changing the taint, can we add a toleration to the kueue pods?

File: modules/management/kubectl-apply/main.tf
Comment (nick-stroud) [inline modules/management/kubectl-apply/main.tf:28]: IIUC, jobset is a dependency for Kueue to work. Why not automatically install jobset if queue is requested. 

```
  install_jobset        = try(var.jobset.install, local.install_kueue)
```

There is precedence for this sort of thing in startup-script module where we automatically install ansible if the user wants to run an ansible script.

Note: I still think there is value to have a separate variable just for jobset (if you only want to install jobset, or control the version)

File: modules/management/kubectl-apply/variables.tf
Comment (nick-stroud) [inline modules/management/kubectl-apply/variables.tf:44]: nit: I am on the fence between picking a known good version as a default or always taking the latest which I assume is what you are doing here.

Comment (chengcongdu) [reply modules/management/kubectl-apply/variables.tf:44]: +1 on stable

File: modules/management/kubectl-apply/variables.tf
Comment (chengcongdu) [inline modules/management/kubectl-apply/variables.tf:45]: are we going to provide a default config in toolkit?

File: examples/gke-a3-highgpu.yaml
Comment (chengcongdu) [reply examples/gke-a3-highgpu.yaml:62]: I would prefer to have kueue system pod in the system NP, mainly for isolation, stability and maintenance consideration.
PREFER_NO_SCHEDULE make it possible for either case, and would be hard to maintain. Toleration for Kueue system pod seems a better fit.

File: modules/management/kubectl-apply/variables.tf
Comment (chengcongdu) [reply modules/management/kubectl-apply/variables.tf:41]: [] () without space

File: examples/gke-a3-highgpu.yaml
Comment (chengcongdu) [reply examples/gke-a3-highgpu.yaml:62]: resourceFlavor might be use for workload taint/toleration setting https://kueue.sigs.k8s.io/docs/reference/kueue.v1beta1/#kueue-x-k8s-io-v1beta1-ResourceFlavorSpec

Comment (ankitkinra) [reply examples/gke-a3-highgpu.yaml:62]: At the moment we are downloading pod installation manifest directly from github without any modifications.
I do not want to modify that file as that has a bunch of things in there, so let me see if we can find if we can apply add tolerations after the manifest has been applied.

File: modules/management/kubectl-apply/variables.tf
Comment (ankitkinra) [reply modules/management/kubectl-apply/variables.tf:44]: I have added the known good version in the local block where I am getting the source url for the download. I would want us not to use the LATEST version without doing some sort of integration test.

File: examples/gke-a3-highgpu.yaml
Comment (ankitkinra) [reply examples/gke-a3-highgpu.yaml:62]: I think there are two things to note:
* kueue-controller-manager which is installed via the downloaded github script, IIUC this doesn't use the ResourceFlavor based tolerations those are for workloads that are being admitted via Kueue
* Kueue workloads which in this case we might not want on the system pod as they would be for the GPU nodepool.

I agree if adding tolerations to `kueue-controller-manager` is possible that is the best way forward.

File: modules/management/kubectl-apply/main.tf
Comment (sharabiani) [inline modules/management/kubectl-apply/main.tf:28]: - AFAIK Jobset is supported in Kueue but it is not a requirement for that.
- However, if we are going to install it with Kueue, it's better not to ignore user's explicit `false` input. Providing false can mean uninstalling Jobset.
- And if we are going to make Kueue's installation depended on Jobset, it's better to control it with precondition.

File: modules/management/kubectl-apply/main.tf
Comment (sharabiani) [inline modules/management/kubectl-apply/main.tf:81]: I'd control if local.install_kueue is true, because it will apply configuration manifest if it's provided but `install` setting is false.

File: modules/management/kubectl-apply/variables.tf
Comment (sharabiani) [reply modules/management/kubectl-apply/variables.tf:44]: Targeting the latest version is tricky because whenever a new version releases deployment will uninstall/install the kueue or jobset for the customers.
I believe changes in the blueprints or upgrading to a new version of gcluster (probably?) should be the expected reasons for Kueue/Jobset updates.

File: examples/gke-a3-highgpu.yaml
Comment (ankitkinra) [reply examples/gke-a3-highgpu.yaml]: Downloaded the manifest and added tolerations

File: modules/management/kubectl-apply/variables.tf
Comment (ankitkinra) [reply modules/management/kubectl-apply/variables.tf]: It is a pinned version and as we would need to update the tolerations we can only accept the versions of kueue/jobset that we have modified, hence added a known version validation

Comment (ankitkinra) [reply modules/management/kubectl-apply/variables.tf]: +1 on not picking LATEST by default.

File: modules/management/kubectl-apply/main.tf
Comment (ankitkinra) [reply modules/management/kubectl-apply/main.tf]: Done, PTAL

File: modules/management/kubectl-apply/main.tf
Comment (ankitkinra) [reply modules/management/kubectl-apply/main.tf]: In order to unblock this PR, I have made kueue and jobset installation independent. We can do this improvement in the next iteration.

File: modules/management/kubectl-apply/main.tf
Comment (ankitkinra) [reply modules/management/kubectl-apply/main.tf]: Done

File: modules/management/kubectl-apply/manifests/kueue-v0.8.1.yaml
Comment (chengcongdu) [inline modules/management/kubectl-apply/manifests/kueue-v0.8.1.yaml:1]: I thought we will be pulling in the config like [this](https://kueue.sigs.k8s.io/docs/installation/#install-a-custom-configured-released-version), and have some logic to add the toleration. Instead of maintaining a local version, is this feasible?
btw, the PR for adding generic toleration to Kueue got turn down. The SIG laying toward letting user to customize the config after ward instead of providing a generic toleration

File: examples/gke-a3-highgpu.yaml
Comment (chengcongdu) [inline examples/gke-a3-highgpu.yaml:68]: please revert this deletion

curious, are you branching from main instead of develop?

File: examples/gke-a3-megagpu.yaml
Comment (chengcongdu) [inline examples/gke-a3-megagpu.yaml:68]: revert deletion

**Review by chengcongdu — APPROVED** [review]
please address the comment before merging

File: modules/management/kubectl-apply/variables.tf
Comment (sharabiani) [inline modules/management/kubectl-apply/variables.tf:51]: nit: Supported **_version_** of Kueue **_is_** [v0.8.1]

File: modules/management/kubectl-apply/variables.tf
Comment (sharabiani) [inline modules/management/kubectl-apply/variables.tf:65]: nit: Supported _**version**_ of Jobset **_is_** [v0.5.2]

File: modules/management/kubectl-apply/main.tf
Comment (sharabiani) [inline modules/management/kubectl-apply/main.tf:61]: nit: maybe we explicitly set `wait_for_rollout = true`? to keep it safe from any changes on default values in kubectl module.

**Review by sharabiani — APPROVED** [review]

**Pull request merged**
